### PR TITLE
Feat: Vcoin Transfer Hook + Claim Contract

### PR DESCRIPTION
**Pull request by SlayerAnsh**
# Pull Request

## Description

- Added vcoin transfer hook msg to trigger executes on vcoin receive. (Currently its limited to native chains only)
- New `claim` contract added that can create claim using vouchers for a pubkey. Users can claim using their pubkey

## Type of Change

- [ ] Bug fix
- [x] New feature
- [ ] Breaking change
- [ ] Documentation update
- [ ] Refactor/Chore

## Related Issue

[Link to related issue, if applicable]

## Testing

Steps to test:

1. [Step 1]
2. [Step 2]
3. [Step 3]

## Checklist

- [ ] Self-review completed
- [ ] Tests added/updated (if applicable)
- [ ] Documentation updated (if applicable)

## Screenshots (if applicable)

[Add screenshots here]

## Additional Notes

[Any additional information or context]

### CODE REVIEW

File: contracts/hub/virtual_balance/src/contract.rs [hunk 1]
```diff
@@ -39,19 +39,19 @@ pub fn instantiate(
...
+    mut deps: DepsMut,
     _env: Env,
     info: MessageInfo,
     msg: ExecuteMsg,
 ) -> Result<Response, ContractError> {
     match msg {
-        ExecuteMsg::Mint(msg) => execute_mint(deps, info, msg),
-        ExecuteMsg::Burn(msg) => execute_burn(deps, info, msg),
-        ExecuteMsg::Transfer(msg) => execute_transfer(deps, info, msg),
+        ExecuteMsg::Mint(msg) => execute_mint(deps.branch(), info, msg),
```
Comment (joemonem) [inline contracts/hub/virtual_balance/src/contract.rs]: the `.branch()` calls seem unnecessary

File: contracts/hub/virtual_balance/src/execute.rs [hunk 2]
```diff
@@ -86,44 +106,70 @@ pub fn execute_burn(
...
-    deps: DepsMut,
+    deps: &mut DepsMut,
     info: MessageInfo,
     msg: ExecuteTransfer,
 ) -> Result<Response, ContractError> {
     let state = STATE.load(deps.storage)?;
 
-    ensure!(!msg.amount.is_zero(), ContractError::ZeroAssetAmount {});
+    let sender = if let Some(sender) = msg.sender {
+        // Only  router can set pseudo sender
```
Comment (joemonem) [inline contracts/hub/virtual_balance/src/execute.rs]: That would be a good error message to have. `Unauthorized: Only router...`

File: contracts/hub/virtual_balance/src/execute.rs [hunk 3]
```diff
@@ -86,44 +106,70 @@ pub fn execute_burn(
 }
 
 pub fn execute_transfer(
-    deps: DepsMut,
+    deps: &mut DepsMut,
     info: MessageInfo,
     msg: ExecuteTransfer,
```
Comment (joemonem) [inline contracts/hub/virtual_balance/src/execute.rs]: I think we should rename this variable to avoid `msg.msg` situations. 
Maybe to `transfer_msg` and within ExecuteTransfer, rename msg to `forward_msg`

File: contracts/hub/virtual_balance/src/execute.rs [hunk 4]
```diff
@@ -135,50 +181,83 @@ pub fn execute_transfer(
...
+    deps: &mut DepsMut,
+    sender: &CrossChainUser,
+    from: &CrossChainUser,
+    amount: Uint128,
+    token_id: &str,
+) -> Result<Vec<Attribute>, ContractError> {
+    let sender_balance_key = BalanceKey {
+        token_id: token_id.to_string(),
+        cross_chain_user: from.clone(),
+    };
```
Comment (joemonem) [inline contracts/hub/virtual_balance/src/execute.rs:237]: we can directly call `to_serialized_balance_key()` here instead of calling it twice below

File: contracts/liquidity/claimer/src/contract.rs [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+use euclid::error::ContractError;
+use euclid::msgs::claimer::{ExecuteMsg, InstantiateMsg, QueryMsg, State};
+use euclid::msgs::factory;
+
+use crate::execute::{execute_claim_voucher, execute_virtual_balance_receive};
+use crate::query::{get_claim, get_sender_claims, get_user_claims};
+use crate::{execute::execute_update_admin, query::get_state, state::STATE};
+
+// version info for migration info
+const CONTRACT_NAME: &str = "crates.io:euclid-relayer";
```
Comment (joemonem) [inline contracts/liquidity/claimer/src/contract.rs]: needs renaming

File: packages/euclid/src/msgs/claimer.rs [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+    GetSenderClaims { sender: String },
+    #[returns(Vec<u128>)]
+    GetUserClaims { pub_key: Binary },
+    #[returns(Claim)]
+    GetClaim { claim_id: u128 },
+}
+
+#[cw_serde]
+pub struct State {
+    // Address of the virtual balance contract
```
Comment (joemonem) [inline packages/euclid/src/msgs/claimer.rs]: confusing comment

File: packages/euclid/src/msgs/factory/msg.rs [hunk 7]
```diff
@@ -63,13 +63,18 @@ pub enum ExecuteMsg {
         token: Token,
         amount: Uint128,
         recipient_address: CrossChainUser,
+        // In user has approval for a transfer, behaves like cw20 allowance
```
Comment (joemonem) [inline packages/euclid/src/msgs/factory/msg.rs]: Confusing comment

File: contracts/liquidity/claimer/src/state.rs [hunk 8]
```diff
@@ -0,0 +1,14 @@
+use cw_storage_plus::{Item, Map};
+use euclid::msgs::claimer::{Claim, State};
+
+pub const STATE: Item<State> = Item::new("state");
+
+pub const CLAIMS: Map<u128, Claim> = Map::new("claims");
```
Comment (joemonem) [inline contracts/liquidity/claimer/src/state.rs:6]: What if we create a struct that contains `Claim`, `Sender`, and `User` as its elements, and use that struct as the value of `CLAIMS`. Would that be enough to no longer need `SENDER_CLAIMS` and `USER_CLAIMS`?

Comment (SlayerAnsh) [reply contracts/liquidity/claimer/src/state.rs:6]: Used this structure for faster key query with two way index, will check if we can update it with the proposed structure.

File: packages/euclid/src/chain.rs [hunk 9]
```diff
@@ -119,6 +119,7 @@ impl CrossChainUser {
         refund_address: Option<String>,
         forwarding_message: Option<EuclidReceive>,
         vcoin_msg: Option<Binary>,
+        unsafe_refund_voucher_to_recipient: Option<bool>,
```
Comment (joemonem) [inline packages/euclid/src/chain.rs:122]: Would be helpful to have a comment that explains this.

File: tests-integration/src/tests/factory.rs [hunk 10]
```diff
@@ -2449,6 +2452,7 @@ fn run_test_stable_pool_swap_request(factory_chain_id: &str, router_chain_id: &s
                     refund_address: None,
                     forwarding_message: None,
                     vcoin_msg: None,
+                    unsafe_refund_voucher_to_recipient: None,
```
Comment (joemonem) [inline tests-integration/src/tests/factory.rs:2461]: Can we add a test case where this variable is set to `true`? Since `None` always defaults to `false`

Comment (SlayerAnsh) [reply tests-integration/src/tests/factory.rs:2461]: Added test for all three cases for withdraw
1. Withdraw success
2. Withdraw fail with unsafe false
3. Withdraw fail with unsafe true

**Review by joemonem — APPROVED** [review]
